cunumeric/deferred.py

```diff
@@ -503,6 +503,25 @@ def _create_indexing_array(
                         f"dimension {i} doesn't match to the shape of the"
                         f"index array which is {rhs.shape[i]}"
                     )
+
+            if key.size == 0:
```

Comment:
This code is never exercised in the current test suite, because it is quite hard to produce 0-size `DeferredArray`s (cunumeric prefers 0-size arrays to be Eager). I wrote some code that thoroughly exercises this part, and it seems to be failing currently:

```
import numpy as np
import cunumeric as cn

from legate.core import LEGATE_MAX_DIM

def mk_deferred_array(lib, shape):
    if np.prod(shape) != 0:
        return lib.ones(shape)
    # for shape (2,0,3,4): good_shape = (2,1,3,4)
    good_shape = tuple(max(1, dim) for dim in shape)
    # for shape (2,0,3,4): key = [:,[False],:,:]
    key = tuple([False] if dim == 0 else slice(None) for dim in shape)
    return lib.ones(good_shape)[key]

for arr_ndim in range(1, LEGATE_MAX_DIM + 1):
    for idx_ndim in range(1, arr_ndim + 1):
        for zero_dim in range(arr_ndim):
            arr_shape = tuple(0 if dim == zero_dim else 3 for dim in range(arr_ndim))
            np_arr = mk_deferred_array(np, arr_shape)
            cn_arr = mk_deferred_array(cn, arr_shape)
            idx_shape = arr_shape[:idx_ndim]
            val_shape = arr_shape if idx_ndim == 1 else (np.prod(idx_shape),) + arr_shape[idx_ndim:]
            np_idx = np.ones(idx_shape, dtype=np.bool_)
            cn_idx = cn.ones(idx_shape, dtype=np.bool_)
            assert np.array_equal(np_arr[np_idx], cn_arr[cn_idx])
            np_val = np.random.random(val_shape)
            cn_val = cn.array(np_val)
            np_arr[np_idx] = np_val
            cn_arr[cn_idx] = cn_val
            assert np.array_equal(np_arr, cn_arr)
```

It currently fails with:

```
[0 - 70000b0c4000]    1.421775 {5}{runtime}: [error 209] LEGION ERROR: The destination indirect field for a copy operation has the incorrect size for the destination region coordinate space. Field 1048595 of destination indirect region requirement 0 is 8 bytes but the coordinate types of the destination space is 16 bytes for copy operation (UID 62) in parent task legion_python_main (UID 1). (from file /Users/mpapadakis/legate.core/legion/runtime/legion/legion_ops.cc:4648)
```

Please make sure it passes, and add it to the test suite.